### PR TITLE
Windows

### PR DESCRIPTION
**Pull request by bam241**
This should allow to build dagmc on windows.

Still a draft as this needs a new version of PyNE. (right now pulling the one from my fork.)

### CODE REVIEW

**Comment by bam241** [comment]
@gonuke @pshriwise this can probably be reviewed while we wait for the  PR pyne/pyne#1367 to be merged.

File: cmake/DAGMC_macros.cmake [hunk 1]
```diff
@@ -35,7 +35,11 @@ macro (dagmc_setup_build)
   execute_process(COMMAND date +%m/%d/%y OUTPUT_VARIABLE ENV_DATE OUTPUT_STRIP_TRAILING_WHITESPACE)
   execute_process(COMMAND date +%H:%M:%S OUTPUT_VARIABLE ENV_TIME OUTPUT_STRIP_TRAILING_WHITESPACE)
 
-  set(CMAKE_STATIC_LIBRARY_SUFFIX ".a")
+  if(MSVC)
```
Comment (gonuke) [inline cmake/DAGMC_macros.cmake]: Hmmm... isn't CMake supposed to get this kind of thing right? I wonder why we set it explicitly in the first place?

File: cmake/FindMOAB.cmake [hunk 2]
```diff
@@ -38,7 +38,10 @@ message(STATUS "HDF5_LIBRARIES_SHARED: ${HDF5_LIBRARIES_SHARED}")
 message(STATUS "HDF5_LIBRARIES_STATIC: ${HDF5_LIBRARIES_STATIC}")
 
 include_directories(${HDF5_INCLUDE_DIRS})
-
+if(MSVC)
+    set(BUILD_STATIC_LIBS TRUE)
+    set(BUILD_SHARED_LIBS OFF)
```
Comment (gonuke) [inline cmake/FindMOAB.cmake:43]: Do shared libs fail, or are they just unnecessary?

File: src/CMakeLists.txt [hunk 3]
```diff
@@ -8,14 +8,14 @@ endif ()
 
 # Required libraries
 add_subdirectory(dagmc)
-add_subdirectory(pyne)
```
Comment (gonuke) [inline src/CMakeLists.txt:11]: Do we only need PyNE for UWUW?

File: src/make_watertight/MakeWatertight.cpp [hunk 4]
```diff
@@ -1752,6 +1752,7 @@ moab::ErrorCode MakeWatertight::seal_surface_loops(
 
   moab::ErrorCode rval;
   for (unsigned j = 0; j < skin.size(); ++j) {
+    skin_loops.push_back(moab::EntityHandle());
```
Comment (gonuke) [inline src/make_watertight/MakeWatertight.cpp]: Can't you `resize()` this before the loop?

File: src/pyne/update_pyne.sh [hunk 5]
```diff
@@ -3,7 +3,7 @@
 set -e
 
 # Clone pyne repo
-git clone --depth 1 https://github.com/pyne/pyne -b develop
+git clone --depth 1 https://github.com/bam241/pyne -b windows_build
```
Comment (gonuke) [inline src/pyne/update_pyne.sh]: This needs to be updated

File: src/uwuw/uwuw.cpp [hunk 6]
```diff
@@ -64,7 +65,11 @@ std::string UWUW::get_full_filepath(std::string filename) {
   filename.erase(std::remove(filename.begin(), filename.end(), ' '),
                  filename.end());
   // use stdlib call
+#ifndef _WIN32
   const char* full_filepath = realpath(filename.c_str(), NULL);
+#else
+  const char* full_filepath = filename.c_str();
```
Comment (gonuke) [inline src/uwuw/uwuw.cpp]: Still interested in getting uniform behavior here...

**Review by gonuke — CHANGES_REQUESTED** [review]
Thanks @bam241 - I have a few questions/comments.

File: cmake/FindMOAB.cmake (hunk 2 above)
Comment (bam241) [reply cmake/FindMOAB.cmake:43]: unnecessary if I recall correctly

File: src/CMakeLists.txt (hunk 3 above)
Comment (bam241) [reply src/CMakeLists.txt:11]: I believe so yes, so we might as well group them together

**Comment by gonuke** [comment]
Looks like this needs a rebase, esp. wrt PyNE

File: src/uwuw/uwuw.cpp (hunk 6 above)
Comment (bam241) [reply src/uwuw/uwuw.cpp]: I still don't know what to do for this one....

Comment (gonuke) [reply src/uwuw/uwuw.cpp]: let's leave it for now

Comment (gonuke) [reply src/uwuw/uwuw.cpp]: Although the simple option is to have all platforms behave like the Windows one.  It is more limiting, but consistent

**Comment by gonuke** [comment]
This looks like it's getting closer.  I wonder if the problem now is that we use forward slash '/' in the file names in `amalgamate_pyne.sh`?

**Comment by bam241** [comment]
I believe the problem occurs when trying to run the bash scripts (on load, before it actually try to run what is in it...)

I found some hint of answer Friday, didn't had time to try them yesterday....

File: src/uwuw/uwuw.cpp (hunk 6 above)
Comment (gonuke) [reply src/uwuw/uwuw.cpp]: Other than the problems caused by this in PyNE, wouldn't `filesystem::canonical()` work here?

Comment (bam241) [reply src/uwuw/uwuw.cpp]: yes I think so
was planing on fixing this if we had a fix for the PyNE compilation

**Comment by bam241** [comment]
I have PR both in MOAB and PyNE to fixe this.

https://bitbucket.org/fathomteam/moab/pull-requests/550/fixing-cpp17-compatibilities 
and pyne/pyne#1390

File: .github/workflows/windows.yml [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+        shell: bash -l {0}
+        run: |
+          cd ${HOME_PATH}
+          mkdir build 
+          #git clone https://github.com/bam241/dagmc -b windows
+          cd dagmc
+          git submodule update --init --recursive
+          # cd src/pyne
+          # ./amalgamate_pyne.sh `pwd` `pwd`
+
```
Comment (bam241) [inline .github/workflows/windows.yml]: this has to be cleaned before merge !

File: .github/workflows/windows.yml [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+          # cd ../../../dagmc_build
+          cd ../build
+          ls D:/a/DAGMC/DAGMC/src/pyne
+          #bash D:/a/DAGMC/DAGMC/src/pyne/amalgamate_pyne.sh D:/a/DAGMC/DAGMC/src/pyne D:/a/DAGMC/DAGMC/src/pyne
```
Comment (bam241) [inline .github/workflows/windows.yml]: this need to be clean as well

File: .gitmodules [hunk 9]
```diff
@@ -1,3 +1,3 @@
 [submodule "src/pyne/pyne"]
 	path = src/pyne/pyne
-	url = https://github.com/pyne/pyne
+	url = https://github.com/bam241/pyne
```
Comment (bam241) [inline .gitmodules]: this should be reverted when PyNE PR gets merged pyne/pyne#1390

File: .tmp_circle/config.yml [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+        type: string
+    docker:
+      - image: svalinn/dagmc-ci-ubuntu-18.04-housekeeping:latest
+        auth:
+          username: $DOCKERHUB_USER
+          password: $DOCKERHUB_PASS
+    working_directory: /root/build_dir/DAGMC
+    steps:
+      - checkout
+      - run:
```
Comment (bam241) [inline .tmp_circle/config.yml]: this file should be removed.

File: cmake/DAGMC_macros.cmake [hunk 11]
```diff
@@ -35,7 +35,11 @@ macro (dagmc_setup_build)
...
+    set(CMAKE_STATIC_LIBRARY_SUFFIX ".lib")
+  else()
+    set(CMAKE_STATIC_LIBRARY_SUFFIX ".a")
+  endif()
```
Comment (bam241) [inline cmake/DAGMC_macros.cmake]: this should be removed again....

**Review by bam241 — COMMENTED** [review]
I can't request change (as it is my PR) but all those comments must be addressed before merge

**Comment by bam241** [comment]
I think we are having a problem:
```
bash /root/build_dir/DAGMC/src/pyne/amalgamate_pyne.sh /root/build_dir/DAGMC/src/pyne /root/build_dir/DAGMC/src/pyne
Traceback (most recent call last):
  File "amalgamate.py", line 145, in <module>
    main()
  File "amalgamate.py", line 125, in main
    hdr.append_file(f)
  File "amalgamate.py", line 75, in append_file
    with open(filename, 'rt', encoding='utf-8') as f:
TypeError: 'encoding' is an invalid keyword argument for this function
```

**Comment by gonuke** [comment]
I relaunched these tests to see if it helps... but it may need a rebase

**Review by gonuke — APPROVED** [review]

**Comment by bam241** [comment]
we probably need to wait for  the MOAB 5.3.0 release

File: .github/workflows/windows.yml [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+      - name: Environment Variables
+        shell: bash -l {0}
+        run: |
+          echo "HOME_PATH=$GITHUB_WORKSPACE/.." >> $GITHUB_ENV
+
+      - name: Build MOAB
+        shell: bash -l {0}
+        run: |
+          cd ${HOME_PATH}
+          git clone --branch cpp17 https://bitbucket.org/bam241/moab
```
Comment (bam241) [inline .github/workflows/windows.yml]: this needs to change to moab 5.3.0

**Comment by bam241** [comment]
@gonuke I set MOAB version to 5.3.0.

It is good on my side, let me know what you still think that's good to go  :)

File: .github/workflows/windows.yml [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+  # allows us to run workflows manually
+  workflow_dispatch:
+  pull_request:
+    branches:
+      - develop
+      - master
+  push:
+    branches:
+      - develop
+      - master
```
Comment (gonuke) [inline .github/workflows/windows.yml]: Do we want any path guards here?

File: .github/workflows/windows.yml [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+
+      - name: Conda dependencies
+        shell:  bash -l {0}
+        run: |
+          conda install curl eigen
+
+      - name: HDF5 download
+        shell:  bash -l {0}
+        run: |
+          curl -L https://support.hdfgroup.org/ftp/HDF5/releases/hdf5-1.8/hdf5-1.8.21/bin/hdf5-1.8.21-Std-win7_64-vs14.zip --output hdf5-1.8.21-Std-win7_64-vs14.zip
```
Comment (gonuke) [inline .github/workflows/windows.yml]: Is this the newest version of HDF5 that works?

File: .github/workflows/windows.yml [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+        with:
+          auto-activate-base: true
+          activate-environment: ""
+
+      - name: Conda dependencies
+        shell:  bash -l {0}
+        run: |
+          conda install curl eigen
+
+      - name: HDF5 download
```
Comment (gonuke) [inline .github/workflows/windows.yml]: Can we conda install HDF5 (since we're in a conda environment anyway?)

File: .github/workflows/windows.yml [hunk 16]
```diff
@@ -0,0 +1,107 @@
...
+                -DMOAB_DIR=../moab_install \
+                -DHDF5_ROOT="C:/Program Files/HDF_Group/HDF5/1.8.21" \
+                -DHDF5_hdf5_LIBRARY_RELEASE="C:/Program Files/HDF_Group/HDF5/1.8.21/lib/libhdf5_hl.lib;C:/Program Files/HDF_Group/HDF5/1.8.21/lib/libhdf5.lib;C:/Program Files/HDF_Group/HDF5/1.8.21/lib/libzlib.lib;C:/Program Files/HDF_Group/HDF5/1.8.21/lib/libszip.lib;C:/Program Files/HDF_Group/HDF5/1.8.21/lib/libhdf5_cpp.lib" \
+                -DCMAKE_INSTALL_PREFIX=../dagmc_install/ \
+                -DCMAKE_EXE_LINKER_FLAGS="" \
+                -DCMAKE_MODULE_LINKER_FLAGS="" \
+                -DCMAKE_SHARED_LINKER_FLAGS="" \
+                -DCMAKE_STATIC_LINKER_FLAGS="" \
+                -DCMAKE_BUILD_TYPE=Release
+          cmake --build . --config Release
```
Comment (gonuke) [inline .github/workflows/windows.yml]: can we run tests?

File: src/pyne/remove_unsupported.py [hunk 17]
```diff
@@ -1,8 +1,16 @@
-with open('pyne.cpp', 'r', encoding='utf-8') as reader:
-    lines = reader.readlines()
+import sys
 
-writer = open('pyne.cpp.new', 'w', encoding='utf-8')
+if sys.version > '3':
```
Comment (gonuke) [inline src/pyne/remove_unsupported.py]: is this because windows is stuck with Py 2?

**Review by gonuke — CHANGES_REQUESTED** [review]
A few new comments with fresh eyes and new experiences.

File: src/pyne/remove_unsupported.py (hunk 17 above)
Comment (bam241) [reply src/pyne/remove_unsupported.py]: I don't know, but with the new change I PRed this shouldn't be necessary....

File: .github/workflows/windows.yml (hunk 16 above)
Comment (bam241) [reply .github/workflows/windows.yml]: I don't know.
I think I unsuccessfully tried.
let me try again :)

File: src/pyne/remove_unsupported.py (hunk 17 above)
Comment (gonuke) [reply src/pyne/remove_unsupported.py]: should we try without? or is this necessary for some users stuck on Py2 systems?

**Comment by bam241** [comment]
@gonuke you were correct conda hdf5 works !

File: .github/workflows/windows.yml (hunk 16 above)
Comment (bam241) [reply .github/workflows/windows.yml]: I don't know how to run the test.

make is not available, so I don't know how to run them...

**Comment by bam241** [comment]
I am not sure how to run the test... I think we build them tho.


I'll try using ctest...

File: .github/workflows/windows.yml [hunk 18]
```diff
@@ -0,0 +1,100 @@
...
+          cmake ../dagmc \
+                -G"Visual Studio 16 2019" \
+                -DBUILD_EXE=OFF \
+                -DBUILD_SHARED_LIBS=ON \
+                -DBUILD_STATIC_LIBS=OFF \
+                -DBUILD_TALLY=OFF \
+                -DBUILD_BUILD_OBB=OFF \
+                -DBUILD_UWUW=ON \
+                -DBUILD_MAKE_WATERTIGHT=ON \
+                -DBUILD_TESTS=OFF \
```
Comment (gonuke) [inline .github/workflows/windows.yml]: Need to turn this on?

**Review by gonuke — CHANGES_REQUESTED** [review]
Maybe we need to turn on tests in CMake?

File: .github/workflows/windows.yml (hunk 18 above)
Comment (bam241) [reply .github/workflows/windows.yml]: that would make sense :)

**Comment by gonuke** [comment]
You may need to wget the test files like we do on linux

**Comment by bam241** [comment]
I don't think we download them... at least not all of them...
`ls` shows them in the `bin`/`test` folder:
```
check_watertight.exe
cones.h5m
cyl.h5m
dag_file.h5m
dagmc_pointinvol_test.exe
dagmc_rayfire_test.exe
dagmc_simple_test.exe
dagmc_unit_tests.exe
make_watertight.exe
make_watertight_cone_tests.exe
make_watertight_cylinder_tests.exe
make_watertight_no_curve_sphere_tests.exe
make_watertight_sphere_n_box_test.exe
mat_lib.h5
MOAB.dll
no_curve_sphere.h5m
pt_vol_test.exe
ray_fire_test.exe
sphere_n_box.h5m
test_dagmc.h5m
test_dagmc_impl.h5m
test_geom.exe
test_geom.h5m
test_uwuw.h5m
uwuw_preproc.exe
uwuw_unit_tests.exe
uwuw_unit_tests_preprocessor.exe
uwuw_unit_tests_tally.exe
```

**Comment by gonuke** [comment]
The Linux install script has a wget line that the windows stuff doesn't include

**Comment by bam241** [comment]
@gonuke I finally got that working :)

**Comment by gonuke** [comment]
Is it necessary to include all of gtest in the PR? or can we install it from source somehow?

File: cmake/DAGMC_macros.cmake [hunk 19]
```diff
@@ -110,7 +109,13 @@ endmacro ()
 macro (dagmc_setup_flags)
   message("")
 
-  set(CMAKE_CXX_STANDARD 11)
+  if(MSVC)
+    set(CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS TRUE)
+    set(CMAKE_CXX_STANDARD 17)
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /std:c++17")
+  else()
+    set(CMAKE_CXX_STANDARD 11)
```
Comment (gonuke) [inline cmake/DAGMC_macros.cmake]: Can't we use C++17 on Linux?

**Comment by bam241** [comment]
it is not new, it was like that before, I just need from `1.7.0` to `1.8.0`

File: cmake/DAGMC_macros.cmake [hunk 20]
```diff
@@ -227,6 +236,8 @@ macro (dagmc_install_library lib_name)
       set_target_properties(${lib_name}-shared
         PROPERTIES INSTALL_RPATH "${INSTALL_RPATH_DIRS}"
                    INSTALL_RPATH_USE_LINK_PATH TRUE)
+
```
Comment (gonuke) [inline cmake/DAGMC_macros.cmake:236]: Cleanup white space?

File: cmake/test_config/CMakeLists.txt [hunk 21]
```diff
@@ -2,7 +2,14 @@ project(Test_DAGMC_config)
...
 
 set(CMAKE_MODULE_PATH ${CMAKE_CURRENT_SOURCE_DIR})
-set(CMAKE_CXX_STANDARD 11)
+
+if(MSVC)
+    set(CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS TRUE)
+    set(CMAKE_CXX_STANDARD 17) 
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /std:c++17")
+else()
+    set(CMAKE_CXX_STANDARD 11)
```
Comment (gonuke) [inline cmake/test_config/CMakeLists.txt]: See above comment

**Comment by gonuke** [comment]
I just realized that this is just an upgrade!

File: .github/workflows/windows.yml [hunk 22]
```diff
@@ -0,0 +1,113 @@
...
+                -DCMAKE_SHARED_LINKER_FLAGS="" \
+                -DCMAKE_STATIC_LINKER_FLAGS="" \
+                -DCMAKE_BUILD_TYPE=Release
+          cmake --build . --config Release
+          cmake --install . --config Release
+
+      - name: test DAGMC
+        shell: bash -l {0}
+        run: |
+          cp -r D:/a/DAGMC/moab_install/* D:/a/DAGMC/dagmc_install/
```
Comment (gonuke) [inline .github/workflows/windows.yml]: Does this copy in all the MOAB libs & includes? If so, are we faking the tests? Should we be able to build against MOAB where it is installed?

**Review by gonuke — CHANGES_REQUESTED** [review]
It's great that we got a working version... now to refine it.

File: .github/workflows/windows.yml (hunk 22 above)
Comment (bam241) [reply .github/workflows/windows.yml]: There is multiple thing there, but you have point:
- I took what I had for the plugin which build every library separately and then combines them into the plugin
- we need a default install folder (like `~/.local` on linux) and add it to the `windows` path

basically I don't think we are faking the test but sloppy/dirty :) 
I'll work on cleaning this up

**Comment by gonuke** [comment]
Also - can we upgrade gtest as a separate PR, first to confine these changes to the Windows stuff?

**Comment by gonuke** [comment]
> Also - can we upgrade gtest as a separate PR, first to confine these changes to the Windows stuff?

And do it as a delete & add rather than a mv

**Comment by gonuke** [comment]
Rebase for #778

**Comment by gonuke** [comment]
How did this break? 😢

Was it moving the gtest into a separate PR?

**Comment by bam241** [comment]
No basically I can get the right way to had the test into the PATH

B. Mouginot

**Comment by gonuke** [comment]
This latest update didn't seem to launch actions from your `windows` branch

**Comment by gonuke** [comment]
Huh - I think it's because you don't launch on push for this workflow.

**Comment by bam241** [comment]
@gonuke I fixed it !

File: .github/workflows/windows.yml [hunk 23]
```diff
@@ -0,0 +1,114 @@
...
+                -DCMAKE_SHARED_LINKER_FLAGS="" \
+                -DCMAKE_STATIC_LINKER_FLAGS="" \
+                -DCMAKE_BUILD_TYPE=Release
+          cmake --build . --config Release
+          cmake --install . --config Release
+
+      - name: Installation Environment Variables
+        shell: bash -l {0}
+        run: |
+          echo "PATH=$PATH:/d/a/DAGMC/DAGMC/../install_dir/bin:/d/a/DAGMC/DAGMC/../install_dir/tests" >> $GITHUB_ENV
```
Comment (gonuke) [inline .github/workflows/windows.yml]: Isn't this the same?

```suggestion
          echo "PATH=$PATH:/d/a/DAGMC/install_dir/bin:/d/a/DAGMC/install_dir/tests" >> $GITHUB_ENV
```

**Review by gonuke — CHANGES_REQUESTED** [review]
Great to finally figure this out!  Just one little question about that path...

File: .github/workflows/windows.yml (hunk 23 above)
Comment (bam241) [reply .github/workflows/windows.yml]: yes ... 
image

**Comment by gonuke** [comment]
One last cleanup request.... rename the tasks in a consistent way:
- "Standard Build and Test / BuildTest" -> "Linux Build and Test / BuildTest"
- "Windows / main" -> "Windows Build and Test / BuildTest"

File: .github/workflows/windows_build_test.yml [hunk 24]
```diff
@@ -0,0 +1,114 @@
...
+  pull_request:
+    branches:
+      - develop
+      - master
+
+env:
+  GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+
+jobs:
+  main:
```
Comment (gonuke) [inline .github/workflows/windows_build_test.yml]:
```suggestion
  BuildTest:
```

**Review by gonuke — CHANGES_REQUESTED** [review]
One last name change

**Comment by gonuke** [comment]
I'm not sure whether or not this needs a rebase?

**Comment by bam241** [comment]
I don't think so, there is no conflict.
there is 1 overlapping change, but they were the exact same... (a missing `#include <array>` ... )

**Comment by gonuke** [comment]
This is such a great step forward! Thanks for tracking this all down @bam241
